### DISCUSSION TITLE
Optimizing group bys with HAVING + LIMIT clauses

### DISCUSSION THREAD

**Discussion by adam-mrozik** [ClickHouse/ClickHouse#67120] [Q&A]
Hey!

I am facing a scenario of high cardinality groupbys (~ 3 billion rows, 1 billion keys).

For 2-3 different columns I need to aggregate results and return 100. Let's imagine a query:

```
SELECT
       key,
       argMax(valueINeed, timestamp) as latest_value
FROM my_table
GROUP BY key
HAVING latest_value > 20
LIMIT 100
```

Now, the query can get VERY slow because of the amount of data. I am currently limiting it with:
```
max_rows_to_group_by=1000000, group_by_overflow_mode='any'
```

But from what I understand, it picks random keys. So If only 0.1% of my keys matches the heuristic of latest_value > 20, it might not find those.

What would be ideal is that Clickhouse somehow, in memory, kept dynamically checking if aggregated keys match the condition of `HAVING` and stop aggregating when it already collects 100 (not sure if it is possible, I am not sure how aggregation algorithm works underneath)

Is something like that achievable? If not, then how would you optimize for a solution like this?

Note that in my case the query is dynamic and  can have ~ 150 different fields with various conditions in HAVING clause, so MV + Aggregate table is not easily achievable, because I would have to aggregate all 150 columns in different ways

**Comment by den-crane** [comment]
Can you share ORDERBY of my_table ?
Maybe you can use  `optimize_aggregation_in_order arg     Enable GROUP BY optimization for aggregating data in corresponding order in MergeTree tables.`

`optimize_aggregation_in_order = 1`

**Reply by adam-mrozik** [reply]
So, I tried to solve it with a smart solution like versioning. I merge each incoming row in redis with its previous state based on `key` column and increment version + 1. I hoped with this I could sort by version desc and just pick latest versions without grouping by. I just forgot about filtering and the fact that `where xxx` in this case could pick up old versions, as they are still in the table : D


So my current order by is:
version (can get max 5)
user_id
ToStartOfHour(timestamp)
key 

EDIT 

but I am open to any other suggestions here, if it would help :) Please note that my scenario will:
always have where filters with timestamp and user_id

always needs latest state of ~ 5 columns based on grouped by key

Can have arbitrary filters from 150 other columns that should be based on the grouped by event (usually something like `HAVING argMax(country, timestamp) = 'US'`)

**Reply by adam-mrozik** [reply]
just tried above setting and still OOM after ~ 15 seconds.

Maybe it would be better if `key` was first, but it surely would mess up speed of fetching data for a specific user/timestamp

**Reply by UnamedRus** [reply]
> But from what I understand, it picks random keys. So If only 0.1% of my keys matches the heuristic of latest_value > 20, it might not find those.


If your WHERE filters are really selective, you can first try to get all keys which match (even in past versions)

And then run query again only for those keys to get latest state and check latest again.
Order by could be something like:

```
ToStartOfHour(timestamp),
user_id,
key,
version (can get max 5)
``` 

OR 

```
user_id,
ToStartOfHour(timestamp),
key,
version (can get max 5)
```

**Reply by adam-mrozik** [reply]
For some users they are not very selective unfortunately, I still end up with billions of events, so it takes a lot of time
